handler/src/test/java/io/netty/handler/timeout/ReadAfterWriteTimeoutHandlerTest.java

```diff
@@ -0,0 +1,360 @@
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.handler.timeout;
+
+import io.netty.bootstrap.Bootstrap;
+import io.netty.bootstrap.ServerBootstrap;
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.Unpooled;
+import io.netty.channel.Channel;
+import io.netty.channel.ChannelHandler;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelInboundHandler;
+import io.netty.channel.ChannelInboundHandlerAdapter;
+import io.netty.channel.ChannelInitializer;
+import io.netty.channel.ChannelOption;
+import io.netty.channel.ChannelPipeline;
+import io.netty.channel.EventLoopGroup;
+import io.netty.channel.nio.NioEventLoopGroup;
+import io.netty.channel.socket.nio.NioServerSocketChannel;
+import io.netty.channel.socket.nio.NioSocketChannel;
+import org.junit.AfterClass;
+import org.junit.BeforeClass;
+import org.junit.Test;
+
+import java.io.IOException;
+import java.net.SocketAddress;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertTrue;
+import static org.junit.Assert.fail;
+
+public class ReadAfterWriteTimeoutHandlerTest {
+    private static EventLoopGroup GROUP;
+
+    /**
+     * We should get {@link io.netty.handler.timeout.ReadTimeoutException} after specified timeout value if server
+     * response with larger latency than the specified timeout.
+     */
+    @Test
+    public void testTrivialTimeoutCase() throws Exception {
+        final int TIMEOUT_MS = 300;
+        final CountDownLatch timeoutGotLatch = new CountDownLatch(1);
+        final CountDownLatch clientClosedChannelLatch = new CountDownLatch(1);
+        final AtomicBoolean unexpectedExceptionInServerHandler = new AtomicBoolean();
+        final AtomicBoolean unexpectedExceptionInClientHandler = new AtomicBoolean();
+
+        ReadAfterWriteTimeoutHandler readAfterWriteTimeoutHandler =
+                new ReadAfterWriteTimeoutHandler(TIMEOUT_MS, TimeUnit.MILLISECONDS);
+
+        ChannelInboundHandler pongHandler = new ChannelInboundHandlerAdapter() {
+            public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception {
+                Thread.sleep(2 * TIMEOUT_MS);
```

Comment:
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png) Remove this use of "Thread.sleep()". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS2925)